edx_rest_framework_extensions/auth/jwt/middleware.py

```diff
@@ -72,19 +86,87 @@ def _add_missing_jwt_permission_classes(self, view_class):
+    a 401 for unauthorized users, but this middleware allows a DRF endpoint
+    to redirect to login.
+
+    This can be used to convert a plan Django view using @login_required into a
+    DRF APIView, which is useful to enable our DRF JwtAuthentication class.
+
+    Usage Notes:
+    - This middleware must be added before JwtAuthCookieMiddleware.
+    - Only affects endpoints using the LoginRequired permission class.
+
+    See https://github.com/edx/edx-platform/blob/master/openedx/core/djangoapps/oauth_dispatch/docs/decisions/0009-jwt-in-session-cookie.rst  # noqa
+    """
+    _LOGIN_REQUIRED_FOUND_CACHE_KEY = 'login_required_found'
+
+    def _get_request_cache(cls):
+        cache_namespace = 'JwtAuthWithLoginRequiredMiddleware'
+        return RequestCache(cache_namespace).data
+
+    def _is_login_required_found(self):
+        """
+        Returns True if LoginRequired permission was found, and False otherwise.
+        """
+        return self._get_request_cache().get(self._LOGIN_REQUIRED_FOUND_CACHE_KEY, False)
+
+    def _check_and_cache_login_required_found(self, view_func):
+        """
+        Checks for LoginRequired permission and caches the result.
+        """
+        view_class = _get_view_class(view_func)
+        view_permission_classes = getattr(view_class, 'permission_classes', tuple())
+        is_login_required_found = _includes_base_class(view_permission_classes, LoginRequired)
+        self._get_request_cache()[self._LOGIN_REQUIRED_FOUND_CACHE_KEY] = is_login_required_found
+
+    def get_login_url(self, request):  # pylint: disable=unused-argument
+        """
+        Return None for default login url.
+
+        Can be overridden for slow-rollout or A/B testing of transition to other login mechanisms.
+        """
+        return None
+
+    def is_jwt_auth_enabled_with_login_required(self, request, view_func):  # pylint: disable=unused-argument
+        """
+        Returns True if JwtAuthentication is enabled with the LoginRequired permission class.
+
+        Can be overridden for slow roll-out or A/B testing.
+        """
+        return self._is_login_required_found()
+
+    def process_view(self, request, view_func, view_args, view_kwargs):  # pylint: disable=unused-argument
+        """
+        Enables Jwt Authentication for endpoints using the LoginRequired permission class.
+        """
+        self._check_and_cache_login_required_found(view_func)
+        if self.is_jwt_auth_enabled_with_login_required(request, view_func):
+            request.META[USE_JWT_COOKIE_HEADER] = 'true'
+
+    def process_response(self, request, response):
+        """
+        Redirects unauthenticated users to login when LoginRequired permission class was used.
+        """
+        if self._is_login_required_found() and not request.user.is_authenticated:
+            login_url = self.get_login_url(request)
+            return login_required(function=lambda request: None, login_url=login_url)(request)
```

Comment:
Nice find in reusing the `login_required` Django function.  BTW, the existence of this function illustrates that Django supports redirecting to login.

Reply:
Django definitely does.  DRF does not. :)

Reply:
We've rephrased things, so I think we are all set.